From 3b201c4864162d78ccc6dd8a61573ff7eff4dccd Mon Sep 17 00:00:00 2001
From: Rameshkumar Sundaram <rameshkumar.sundaram@oss.qualcomm.com>
Date: Wed, 1 Jul 2026 12:27:12 +0530
Subject: [PATCH] Fix incorrect iface removal during config reload for AP MLD

hostapd_reload_config() currently removes interfaces by matching the
interface name. In AP MLD configurations, multiple links can share the
same interface name, which can result in removing the wrong interface
during reload. For example, reloading the 2.4 GHz link may incorrectly
remove the 6 GHz interface.

Example:
  iface[0]->bss[0] => wlan0.0 => 6 GHz
  iface[1]->bss[0] => wlan0.0 => 5 GHz
  iface[2]->bss[0] => wlan0.0 => 2.4 GHz

Fix this by introducing hostapd_remove_hapd_iface(), which removes an
interface by pointer comparison instead of name matching, and use it in
hostapd_reload_config(). Also refactor hostapd_remove_iface() to reuse
this helper and avoid code duplication.

Signed-off-by: Rameshkumar Sundaram <rameshkumar.sundaram@oss.qualcomm.com>
---
 src/ap/hostapd.c | 57 +++++++++++++++++++++++++++++++++---------------
 1 file changed, 40 insertions(+), 17 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index b86553ece..8c88d023b 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -66,6 +66,7 @@ static int hostapd_broadcast_wep_clear(struct hostapd_data *hapd);
 #endif /* CONFIG_WEP */
 static int setup_interface2(struct hostapd_iface *iface);
 static void channel_list_update_timeout(void *eloop_ctx, void *timeout_ctx);
+static void hostapd_cleanup_unused_mlds(struct hapd_interfaces *interfaces);
 static void hostapd_interface_setup_failure_handler(void *eloop_ctx,
 						    void *timeout_ctx);
 
@@ -256,6 +257,37 @@ static int hostapd_iface_conf_changed(struct hostapd_config *newconf,
 }
 
 
+static int hostapd_remove_hapd_iface(struct hostapd_iface *hapd_iface)
+{
+	struct hapd_interfaces *interfaces;
+	size_t i;
+
+	hapd_iface->driver_ap_teardown =
+		!!(hapd_iface->drv_flags &
+		   WPA_DRIVER_FLAGS_AP_TEARDOWN_SUPPORT);
+
+	interfaces = hapd_iface->interfaces;
+	for (i = 0; i < interfaces->count; i++) {
+		if (interfaces->iface[i] == hapd_iface)
+			break;
+	}
+
+	if (i >= interfaces->count)
+		return -1;
+
+	hostapd_interface_deinit_free(hapd_iface);
+
+	while (i < interfaces->count - 1) {
+		interfaces->iface[i] = interfaces->iface[i + 1];
+		i++;
+	}
+	interfaces->count--;
+
+	hostapd_cleanup_unused_mlds(interfaces);
+	return 0;
+}
+
+
 int hostapd_reload_config(struct hostapd_iface *iface)
 {
 	struct hapd_interfaces *interfaces = iface->interfaces;
@@ -292,7 +324,12 @@ int hostapd_reload_config(struct hostapd_iface *iface)
 			hostapd_config_free(newconf);
 			return -1;
 		}
-		hostapd_remove_iface(interfaces, hapd->conf->iface);
+		if (hostapd_remove_hapd_iface(iface) != 0) {
+			os_free(fname);
+			hostapd_config_free(newconf);
+			return -1;
+		}
+
 		iface = hostapd_init(interfaces, fname);
 		os_free(fname);
 		hostapd_config_free(newconf);
@@ -4075,7 +4112,7 @@ static int hostapd_remove_bss(struct hostapd_iface *iface, unsigned int idx)
 int hostapd_remove_iface(struct hapd_interfaces *interfaces, char *buf)
 {
 	struct hostapd_iface *hapd_iface;
-	size_t i, j, k = 0;
+	size_t i, j;
 
 	for (i = 0; i < interfaces->count; i++) {
 		hapd_iface = interfaces->iface[i];
@@ -4083,21 +4120,7 @@ int hostapd_remove_iface(struct hapd_interfaces *interfaces, char *buf)
 			return -1;
 		if (!os_strcmp(hapd_iface->conf->bss[0]->iface, buf)) {
 			wpa_printf(MSG_INFO, "Remove interface '%s'", buf);
-			hapd_iface->driver_ap_teardown =
-				!!(hapd_iface->drv_flags &
-				   WPA_DRIVER_FLAGS_AP_TEARDOWN_SUPPORT);
-
-			hostapd_interface_deinit_free(hapd_iface);
-			k = i;
-			while (k < (interfaces->count - 1)) {
-				interfaces->iface[k] =
-					interfaces->iface[k + 1];
-				k++;
-			}
-			interfaces->count--;
-			hostapd_cleanup_unused_mlds(interfaces);
-
-			return 0;
+			return hostapd_remove_hapd_iface(hapd_iface);
 		}
 
 		for (j = 0; j < hapd_iface->conf->num_bss; j++) {
-- 
2.55.0
